alerting/alerter-war/src/main/java/org/hawkular/metrics/alerter/groups/GroupTriggerManager.java

```diff
@@ -0,0 +1,597 @@
+
+        mgtExecutor = new ScheduledThreadPoolExecutor(THREAD_POOL_SIZE);
+
+        refresh();
+
+        if (null == definitionsListener) {
+            log.info("Registering Trigger UPDATE/REMOVE listener");
+            definitionsListener = new DefinitionsListener() {
+                @Override
+                public void onChange(Set<DefinitionsEvent> event) {
+                    if (coordinator) {
+                        log.warnf("Refreshing due to change event %s", event);
+                        refresh();
+                    }
+                }
+            };
+            definitions.registerListener(definitionsListener, DefinitionsEvent.Type.TRIGGER_CONDITION_CHANGE,
+                    DefinitionsEvent.Type.TRIGGER_REMOVE);
+        }
+    }
+
+    @PreDestroy
+    public void shutdown() {
+        if (coordinator) {
+            stop();
+        }
+
+        if (distributed) {
+            cacheManager.removeListener(topologyChangeListener);
+            cacheManager.stop();
+        }
+    }
+
+    public void stop() {
+        log.infof("Stopping Hawkular Metrics Group Trigger Manager, distributed=%s", distributed);
+
+        if (null != mgtFutures) {
+            mgtFutures.values().forEach(f -> f.cancel(true));
+        }
+        if (null != mgtExecutor) {
+            mgtExecutor.shutdown();
+            mgtExecutor = null;
+        }
+    }
+
+    public void cancel(Trigger mgt) {
+        ScheduledFuture<?> f = mgtFutures.get(mgt);
+        if (null != f) {
+            log.warnf("Canceling Hawkular Metrics Group Trigger Job for %s", mgt);
+            try {
+                f.cancel(true);
+                mgtFutures.remove(mgt);
+            } catch (Exception e) {
+                log.warnf("Failed to cancel Hawkular Metrics Group Trigger Job for %s: %s", mgt, e.getMessage());
+            }
+        }
+    }
+
+    private void refresh() {
+        try {
+            log.warn("Refreshing Metrics Group Triggers!");
+
+            // get all of the metrics group triggers (tagged for hawkular metrics)
+            Collection<Trigger> taggedMgts = definitions.getAllTriggersByTag(
+                    TAG_GROUP_TRIGGER,
+                    TAG_GROUP_TRIGGER_VALUE);
+            log.warnf("Found [%s] Metrics Group Triggers!", taggedMgts.size());
+
+            // generate valid set of metrics group trigger ids
+            Set<Trigger> mgts = taggedMgts.stream()
+                    .filter(t -> {
+                        if (!t.isGroup()) {
+                            log.warnf("Tagged as Metrics Group Trigger is not a group trigger, skipping %s", t);
+                            return false;
+                        }
+                        return true;
+                    })
+                    .collect(Collectors.toSet());
+
+            // for each metrics group trigger run a job to manage its member trigger set
+            for (Trigger mgt : mgts) {
+                if (mgtFutures.containsKey(mgt)) {
```

Comment:
>  but what happen if you want to modify an existing trigger

Right, good question.  There are two things going on.  The change listener is responsible for identifying the set of Metrics Group Triggers (MGTs) and ensuring that each has a recurring job running.  Each MGT job is responsible for maintaining the set of member triggers for that MGT.  So, if an MGT comes or goes the change listener will add/remove a job.  If the MGT is updated with different tag values, or if its metric set changes,  the member set will be updated to reflect those changes.   If this makes it more clear let me know and I'll add this blurb to the class doc.